### ISSUE TITLE
Problems custom fields & Elastic 7.3.2

### ISSUE THREAD

**Issue by ipworkx** [austin-taylor/flare#38]
Hi,
Great tool but I experience some errors during execution. Could it be possible that is has something to do related to Elastic 7.3.2? 
Here is my elasticsearch.ini
[beacon]
es_host=172.16.32.5
es_index=suricata-*
es_port=9200
es_timeout=480
min_occur=10
min_interval=10
min_percent=2
window=30
threads=2
period=2
kibana_version=5
verbose=True
debug=True

#Elasticsearch fields for beaconing
field_source_ip=src_ip
field_destination_ip=dst_ip
field_destination_port=dst_port
field_timestamp=@timestamp
field_flow_bytes_toserver=flow.bytes_toserver
field_flow_id=flow_id
#If using bro, may need to replace with conn
event_type=suricata

#Authentication
username=''
password=''

#set to false if using suricata defaults if you have custom fields
suricata_defaults = false
~                                                                                                                                             
~                                                      
I turned on debugging and this is what I got:

# flare_beacon --verbose --group --whois --focus_outbound -c /home/user/flare/configs/elasticsearch.ini -json beacon.json
[INFO] Attempting to connect to elasticsearch...
[SUCCESS] Connected to elasticsearch on 172.16.32.5:9200
[INFO] Gathering flow data... this may take a while...
[DEBUG] {'query': {'bool': {'filter': [{'bool': {'must_not': [], 'must': [{'range': {'@timestamp': {'gte': 1569262862007, 'lte': 1569270062007, 'format': 'epoch_millis'}}}]}}, {'term': {'event.type': 'suricata'}}], 'must': {'query_string': {'query': '*', 'analyze_wildcard': 'true'}}}}, '_source': ['src_ip', 'dst_ip', 'dst_port', '@timestamp', 'flow.bytes_toserver', 'flow_id']}
[DEBUG] {'query': {'bool': {'filter': [{'bool': {'must_not': [], 'must': [{'range': {'@timestamp': {'gte': 1569262862007, 'lte': 1569270062007, 'format': 'epoch_millis'}}}]}}, {'term': {'event.type': 'suricata'}}], 'must': {'query_string': {'query': '*', 'analyze_wildcard': 'true'}}}}, '_source': ['src_ip', 'dst_ip', 'dst_port', '@timestamp', 'flow.bytes_toserver', 'flow_id']}
[DEBUG]                      @timestamp           dst_ip  dst_port                       flow       flow_id         src_ip
0      2019-09-23T18:22:47.707Z              NaN       NaN                        NaN           NaN            NaN
1      2019-09-23T18:22:27.671Z              NaN       NaN                        NaN           NaN            NaN
2      2019-09-23T18:23:32.295Z              NaN       NaN                        NaN           NaN            NaN
3      2019-09-23T18:22:07.592Z              NaN       NaN                        NaN           NaN            NaN
4      2019-09-23T18:23:02.180Z              NaN       NaN                        NaN           NaN            NaN
5      2019-09-23T18:21:21.954Z              NaN       NaN                        NaN           NaN            NaN
6      2019-09-23T18:28:38.549Z              NaN       NaN                        NaN           NaN            NaN
7      2019-09-23T18:28:33.035Z              NaN       NaN                        NaN           NaN            NaN
8      2019-09-23T18:28:08.512Z              NaN       NaN                        NaN           NaN            NaN
9      2019-09-23T18:28:13.021Z              NaN       NaN                        NaN           NaN            NaN
10     2019-09-23T18:30:18.890Z              NaN       NaN                        NaN           NaN            NaN
11     2019-09-23T18:30:08.885Z              NaN       NaN                        NaN           NaN            NaN
12     2019-09-23T18:29:43.262Z              NaN       NaN                        NaN           NaN            NaN
13     2019-09-23T18:27:48.439Z              NaN       NaN                        NaN           NaN            NaN
14     2019-09-23T18:29:53.272Z              NaN       NaN                        NaN           NaN            NaN
15     2019-09-23T18:29:23.190Z              NaN       NaN                        NaN           NaN            NaN
16     2019-09-23T18:29:03.105Z              NaN       NaN                        NaN           NaN            NaN
17     2019-09-23T18:37:50.331Z              NaN       NaN                        NaN           NaN            NaN
18     2019-09-23T18:37:54.840Z              NaN       NaN                        NaN           NaN            NaN
19     2019-09-23T18:37:00.201Z              NaN       NaN                        NaN           NaN            NaN
20     2019-09-23T18:35:30.029Z              NaN       NaN                        NaN           NaN            NaN
21     2019-09-23T18:36:24.585Z              NaN       NaN                        NaN           NaN            NaN
22     2019-09-23T18:34:54.367Z              NaN       NaN                        NaN           NaN            NaN
23     2019-09-23T18:36:40.138Z              NaN       NaN                        NaN           NaN            NaN
24     2019-09-23T18:34:39.942Z              NaN       NaN                        NaN           NaN            NaN
25     2019-09-23T18:36:14.575Z              NaN       NaN                        NaN           NaN            NaN
26     2019-09-23T18:37:14.725Z              NaN       NaN                        NaN           NaN            NaN
27     2019-09-23T18:43:01.260Z              NaN       NaN                        NaN           NaN            NaN
28     2019-09-23T18:43:05.915Z              NaN       NaN                        NaN           NaN            NaN
29     2019-09-23T18:42:41.240Z              NaN       NaN                        NaN           NaN            NaN
...                         ...              ...       ...                        ...           ...            ...
15767  2019-09-23T20:20:39.734Z    192.168.12.65   52346.0                        NaN  1.471021e+15   192.168.12.1
15768  2019-09-23T20:19:23.909Z     192.168.12.1      53.0                        NaN  1.504029e+15  192.168.12.20
15769  2019-09-23T20:20:40.477Z     192.168.12.1      53.0                        NaN  1.517610e+14  192.168.12.20
15770  2019-09-23T20:19:26.224Z     192.168.12.1      53.0                        NaN  1.824670e+15  192.168.12.20
15771  2019-09-23T20:20:40.489Z    192.168.12.20   51121.0                        NaN  1.517610e+14   192.168.12.1
15772  2019-09-23T20:19:22.384Z    192.168.12.20   42863.0                        NaN  5.343826e+14   192.168.12.1
15773  2019-09-23T20:20:42.023Z     192.168.11.1      53.0    {u'bytes_toserver': 83}  1.947489e+15  192.168.11.30
15774  2019-09-23T20:20:42.023Z     192.168.11.1      53.0    {u'bytes_toserver': 89}  2.248017e+15  192.168.11.30
15775  2019-09-23T20:19:26.237Z    192.168.12.20   46984.0                        NaN  1.824670e+15   192.168.12.1
15776  2019-09-23T20:20:41.090Z      172.16.32.2     968.0                        NaN  1.116387e+15  192.168.11.10
15777  2019-09-23T20:19:56.191Z      172.16.32.6    1015.0                        NaN  1.859919e+15  192.168.11.10
15778  2019-09-23T20:20:48.013Z     192.168.11.1      53.0   {u'bytes_toserver': 167}  2.018723e+15  192.168.11.62
15779  2019-09-23T20:19:59.024Z    192.168.12.20   51729.0                        NaN  1.504029e+15   192.168.12.1
15780  2019-09-23T20:20:45.886Z    192.168.12.20   42408.0                        NaN  2.010799e+15   192.168.12.1
15781  2019-09-23T20:20:42.022Z     192.168.11.1      53.0    {u'bytes_toserver': 91}  7.555262e+14  192.168.11.30
15782  2019-09-23T20:20:51.080Z     192.168.11.1      53.0    {u'bytes_toserver': 74}  1.253790e+15  192.168.11.62
15783  2019-09-23T20:20:46.022Z   184.50.162.217     443.0  {u'bytes_toserver': 3958}  1.482827e+15  192.168.11.62
15784  2019-09-23T20:20:49.801Z    192.168.12.65   52364.0                        NaN  1.964770e+15   192.168.12.1
15785  2019-09-23T20:20:47.014Z   139.162.192.13     123.0    {u'bytes_toserver': 90}  1.375954e+15  192.168.11.71
15786  2019-09-23T20:20:48.013Z     192.168.11.1      53.0   {u'bytes_toserver': 144}  7.315431e+14  192.168.11.62
15787  2019-09-23T20:20:48.013Z     192.168.11.1      53.0    {u'bytes_toserver': 97}  7.970951e+14  192.168.11.62
15788  2019-09-23T20:20:44.758Z    192.168.12.65   52353.0                        NaN  1.565523e+15   192.168.12.1
15789  2019-09-23T20:20:45.885Z     192.168.12.1      53.0                        NaN  2.010799e+15  192.168.12.20
15790  2019-09-23T20:20:46.145Z     192.168.12.1      53.0                        NaN  7.998153e+13  192.168.12.20
15791  2019-09-23T20:20:46.523Z     192.168.12.1      53.0                        NaN  5.601224e+14  192.168.12.65
15792  2019-09-23T20:20:45.886Z    192.168.12.20   46984.0                        NaN  1.824670e+15   192.168.12.1
15793  2019-09-23T20:20:48.226Z    192.168.12.20   51121.0                        NaN  1.517610e+14   192.168.12.1
15794  2019-09-23T20:20:51.080Z   192.168.11.255     138.0   {u'bytes_toserver': 243}  1.183719e+15  192.168.11.10
15795  2019-09-23T20:20:49.786Z    77.74.177.233      80.0                        NaN  2.027704e+14  192.168.12.65
15796  2019-09-23T20:20:51.014Z  255.255.255.255      67.0   {u'bytes_toserver': 676}  2.410321e+14        0.0.0.0

[15797 rows x 6 columns]
ERROR: 'float' object has no attribute 'get'

Any help is welcome

Regard
Thierry

**Comment by SMAPPER** [comment]
The NaN values means that Elasticsearch is returning records that do not have certain fields. I'm working on an update to only bring back records that contain the fields being analyzed.

My gut is you are analyzing Zeek (formerly Bro) conn logs. In conn logs you may not always have source_ip, destination_ip, and destination_port field/value pairs. This can be because IPv6 is treated differently or records contain other data.

**Comment by SMAPPER** [comment]
I have updated Flare as promised. Can you try again?

**Comment by ipworkx** [comment]
Hi,
Thanks for the fast response. I just tried and found two things:
1. The ini files are missing a new upcoming feature "filter=...." this setting was not in the files, so I added it. 
2. Now it fails to connect to because of a missing module Requests

[INFO] Attempting to connect to elasticsearch...
Please install requests to use RequestsHttpConnection.
ERROR: Could not connect to ElasticSearch -- Please verify your settings are correct and try again.

It was there, but I reinstalled it.
So I installed it like : pip install requests==2.9.1

Now I get the following:
[DEBUG] {'query': {'bool': {'filter': [{'bool': {'must_not': [], 'must': [{'range': {'@timestamp': {'gte': 1569370181306, 'lte': 1569391781306, 'format': 'epoch_millis'}}}]}}, {'term': {'event_type': 'flow'}}], 'must': {'query_string': {'query': '_exists_:src_ip AND _exists_:dest_port AND _exists_:dest_ip', 'analyze_wildcard': 'true'}}}}, '_source': ['src_ip', 'dest_ip', 'dest_port', '@timestamp', 'flow.bytes_toserver', 'flow_id']}
ERROR: ConnectionError([SSL: UNKNOWN_PROTOCOL] unknown protocol (_ssl.c:590)) caused by: SSLError([SSL: UNKNOWN_PROTOCOL] unknown protocol (_ssl.c:590))

Regards
Thierry

**Comment by SMAPPER** [comment]
Can you add the below to your ini and try again?

use_ssl=False

**Comment by ipworkx** [comment]
Hi,

I thought so. Same problem like the debug thing. There's still an issue
there. Anyway, I tried but it failed. same problem. Even when I
modifying by setting false to False. (Found out that the stuff was case
sensitive)

Finally I changed the code to set the "use_ssl=False hardcoded in the
module command_control.py. At the end it results in the same issue
regarding complaining about "ERROR: 'float' object has no attribute 'get'"

Let me explain a little about the environment. I'm not using bro yet. It
is still suricata. But builded from the ground up using the setup of Rob
Cowart (https://github.com/robcowart/synesis_lite_suricata) by using
several inside pfsense firewalls and an outside FPC Suricata sensors.
All instances are sending data by reading the eve.json through filebeat
which sends it to several logstash instances located in the central
environment. After all it ends up in a cluster build up by three Elastic
nodes build on Elastic 7.2.3.

The dashboards and hunting is fine, so I thought bringing in a little
excitement by adding smart technology like Flare. Why Flare? Because it
can create json output as well. The only thing I have to do is changing
some fieldnames in the ini file and there we go. But it didn´t work out
like that.

So I added a SELKS5.0 machine and looked at the differences. WIth
SELKS5.0 it works there fine, but it was based on Elastic 6.8.3 I
believe. The query created by Flare was exact the same as the one on
Elastci 7.2.3.

The only difference where it is all about is that the field
flow.bytes_toserver in the 7.2.3 stack is defined as filedtype=number
and format=bytes and within SELKS it is fieldtype=number and
format=-Default-

I think that's where the problem is. Unfortunately I'm not that deep
into the code to find out where I can solve this one. When I start to
dig, I'm just getting deeper without solving the real problem.

I hope I pointed you to the right way here. It's probably a small thing.

BR,

Thierry

**Comment by ipworkx** [comment]
Hi again,

I changed back to 6.8.4 on an ubuntu machine 16.04 LTS and still have issues regarding SSL stuff. Also when I changed this setting in the INI and by doing it hardcoded. 
When I finally removed it hardcodes I get errors like:
ERROR: TransportError(406, u'Content-Type header [] is not supported')

Are you able to help me out here? It seems like that there was no progress the last two months. Am I right?

Regards,
Thierry
